src/server/frontend_xwayland/xwayland_client_manager.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2020 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <william.wold@canonical.com>
+ */
+
+#ifndef MIR_FRONTEND_XWAYLAND_CLIENT_MANAGER_H
+#define MIR_FRONTEND_XWAYLAND_CLIENT_MANAGER_H
+
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+
+namespace mir
+{
+namespace shell
+{
+class Shell;
+}
+namespace scene
+{
+class Session;
+}
+namespace frontend
+{
+
+class XWaylandClientManager
+{
+public:
+    XWaylandClientManager(std::shared_ptr<shell::Shell> const& shell);
+    ~XWaylandClientManager();
+
+    /// Returns the session associated with the given client_pid, or creates a new one if needed
+    /// Stores a strong reference to the session associated with the given owner_key
+    auto get_session_for_client(void* owner_key, pid_t client_pid) -> std::shared_ptr<scene::Session>;
```

Comment:
1. The comment appears to describe how the function is implemented but does not agree with the function name on what the function is used for.
2. I don't find `owner_key` a particularly helpful name. The only production-code use is with an `XWaylandSurface*`, so `client_surface` would seem to be a clearer name. And is there any advantage to `void*` as opposed to something more specific?
I think the function does two things: it tracks the surfaces associated with a client and returns a session. That would be better named:
```c++
auto register_surface_for_client(pid_t client_pid, XWaylandSurface* client_surface) -> std::shared_ptr<scene::Session>;
```
Or as two functions:
```c++
void register_surface_for_client(pid_t client_pid, XWaylandSurface* client_surface);
auto session_for_client(pid_t client_pid) -> std::shared_ptr<scene::Session>;
```

Reply:
I use a `void*` simply because I can. The `XWaylandClientManager` doesn't do anything with the surface or care that it's a surface, so why should it get a strongly typed pointer?

Reply:
The two function version doesn't make sense, since the caller doesn't know or care if it's the first surface with that PID. Determining that is one of the main jobs of this abstraction.

I would be willing to change the names, but I don't see why we should use a strongly typed pointer when we're not using that pointer as anything but an ID. Weakest coupling possible is generally a best practice.

Reply:
> The two function version doesn't make sense, since the caller doesn't know or care if it's the first surface with that PID. Determining that is one of the main jobs of this abstraction.

The two function version doesn't require the caller to know or care if it's the first surface with that PID, it just separates out the "tell" from the "ask". (https://martinfowler.com/bliki/TellDontAsk.html)

> I don't see why we should use a strongly typed pointer when we're not using that pointer as anything but an ID. Weakest coupling possible is generally a best practice.

Stop thinking about how it is implemented, and think about how it is expected to be used. "Owner" is not part of the domain: anyone reading code using this will have to stop and figure out what an "owner" is. There is a meaningful name already in use, why not use it?